### PR TITLE
dev/core#5053 Afform - Add support for saving event location

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Fulfills feature request from https://lab.civicrm.org/dev/core/-/issues/5053

Before
------
Not possible to save email/phone/address when creating an event via Afform

After
------
Support added for those fields, in the form of a block.

### CODE REVIEW

**Comment by mattwire** [comment]
Test fail is probably related:
```
api\v4\Afform\AfformCustomFieldUsageTest::testMultiRecordCustomBlock
Failed asserting that null matches expected 'One'.

/home/homer/buildkit/build/build-3/web/sites/all/modules/civicrm/ext/afform/mock/tests/phpunit/api/v4/Afform/AfformCustomFieldUsageTest.php:102
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2307
```

**Comment by mattwire** [comment]
Non-blocking comment but probably our only opportunity if we wanted to.. How hard would it be to rename the entity from LocBlock to LocationBlock?

**Comment by colemanw** [comment]
> Non-blocking comment but probably our only opportunity if we wanted to.. How hard would it be to rename the entity from LocBlock to LocationBlock?

The LocBlock entity is a horrible design and the name is the least of its problems. I wouldn't bother renaming it since what it really needs is to be ripped out.

File: ext/afform/core/Civi/Api4/Action/Afform/AbstractProcessor.php [hunk 1]
```diff
@@ -272,30 +272,44 @@ protected static function getJoinWhereClause(FormDataModel $formDataModel, strin
...
+    // Forward FK e.g. Event.loc_block_id => LocBlock
+    $forwardFkField = self::getFkField($mainEntityType, $joinEntityType);
+    if ($forwardFkField) {
+      $joinIdField = $forwardFkField['fk_column'];
+      $mainEntityJoinValue = civicrm_api4($mainEntityType, 'get', [
+        'checkPermissions' => FALSE,
+        'where' => [['id', '=', $mainEntityId]],
+        'select' => [$forwardFkField['name']],
+      ])->first();
+      return [$joinIdField, '=', $mainEntityJoinValue[$forwardFkField['name']] ?? 0];
```
Comment (jitendrapurohit) [inline ext/afform/core/Civi/Api4/Action/Afform/AbstractProcessor.php:284]: @colemanw Should this return an array of current return value? I get the following error when trying to prefill the afform with event id in URL (happens only if location block is included).

>TypeError: array_pad(): Argument #1 ($array) must be of type array, string given in array_pad() (line 303 of web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php).

**Comment by Guydn** [comment]
I could look at the patch this morning. I do not see the feature allowing to retrieve an existing location (like on the backend ).

**Comment by Guydn** [comment]
I have just started to test. On the back end there is a field to state whether it is a new location or not. This field does not exist in FB. Maybe I should add a custom field so that I display the existing location or not (cond rule).

I try to create an autocomplete on the addresses. It works well in SK (with address, postal code, town). In form builder, I can set the Search and the autocomplete display. But when i run the form, it only shows the ID and does not apply the criteria (to limit the number of addresses).

**Comment by colemanw** [comment]
Thanks for testing @Guydn 
I should clarify that this PR exposes the entities to Afform but does not attempt to reproduce all the complex features of the "Location" tab on the "Event" screen.
If you let me know what are the most important features for you I can try to work on them.

**Comment by Guydn** [comment]
Yes sure. In the membership organisation, around 50 people are able to create events (and modify). They do not have access to CiviCRM but to a web logged in environment on WP. They are able to choose a event template (that brings a lot of information) and have to define a location (either select one or create one), description, title, dates, quota. 
I have a quick form which has been working for 2,5 years now so it is not urgent.
The location date structure is complex and honestly, for events, I do not have a clear understanding (we have many duplicates...)
Civicrm event location
CiviCRM event location 2
Hoping it clarifies.

**Comment by colemanw** [comment]
@Guydn I've added support for autocomplete + autofill of location info. It was extremely challenging!

**Comment by Guydn** [comment]
> @Guydn I've added support for autocomplete + autofill of location info. It was extremely challenging!

Thanks a lot @colemanw Yes i understand that it is a very very difficult development.

**Comment by allinappliadmin** [comment]
I can't merge patch on 5.76.1 as a file is missing
```
patching file Civi/Api4/Action/LocBlock/Create.php
patching file Civi/Api4/Action/LocBlock/LocBlockSaveTrait.php
patching file Civi/Api4/Action/LocBlock/Save.php
patching file Civi/Api4/Action/LocBlock/Update.php
patching file Civi/Api4/Generic/Traits/DAOActionTrait.php
patching file Civi/Api4/LocBlock.php
patching file Civi/Api4/Utils/CoreUtil.php
patching file ext/afform/admin/Civi/AfformAdmin/AfformAdminMeta.php
patching file ext/afform/admin/afformEntities/LocBlock.php
patching file ext/afform/admin/ang/afGuiEditor/afGuiEntity.component.js
Hunk #1 succeeded at 104 (offset -4 lines).
patching file ext/afform/admin/ang/afGuiEditor/elements/afGuiContainer.component.js
patching file ext/afform/admin/ang/afGuiEditor/elements/afGuiContainer.html
patching file ext/afform/core/Civi/Afform/Behavior/ContactAutofill.php
can't find file to patch at input line 379
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/ext/afform/core/Civi/Afform/Behavior/GroupSubscription.php b/ext/afform/core/Civi/Afform/Behavior/GroupSubscription.php
|index c6b42fdb601e..557fc01ac721 100644
|--- a/ext/afform/core/Civi/Afform/Behavior/GroupSubscription.php
|+++ b/ext/afform/core/Civi/Afform/Behavior/GroupSubscription.php
--------------------------
File to patch: ext/afform/core/Civi/Afform/Behavior/GroupSubscription.php
ext/afform/core/Civi/Afform/Behavior/GroupSubscription.php: No such file or directory
```

image

**Comment by colemanw** [comment]
@allinappliadmin that file exists in the RC but not your version. I don't think it's critical to this PR so you could just ignore it.

@civicrm-builder retest this please

**Comment by allinappliadmin** [comment]
should the test work on the 5.76.1?

**Comment by colemanw** [comment]
@allinappliadmin I think so. One or more files will not patch, but that's probably ok.

**Comment by Guydn** [comment]
There is now a "block address" that is available in the event fields in FB. There are 3 main display possibilities: "autocompletion"," id. location", "select location form". 

I prepared an autocompletion search (on entity address "with" address addresses "with" event location, "grouped by" id. address). When i select autocompletion, i try to select my search but I cannot see it and therefore cannot see my autocompletion display. I'm missing something.
There is a possibility to select a quick add but none are available and I do not see how I can create one.
Actually the global idea is very interesting :-): search in a list of addresses, select one and keep the option to create a new one through a quick add.

If I select "id. location", it is an integer. I do not see how this option can be used.

If i select "select location form", i do not understand how to use it. If I want to customize the options then the whole FB edit screen is frozen.

**Comment by Guydn** [comment]
Yes I get directly the french translation ...
Event location 3
Event location 4


The Id location
Event location 2

The quick add
Event location 1

The select location form
Event location 6

**Comment by colemanw** [comment]
@Guydn ok, I see the confusion. That field is "address_id" and it should not be there. I'm removing it. Sorry for the confusion.
The field you want is "Existing Location" - it is the first field in the list. Dragging a location block onto the form should now look like this:

image

**Comment by allinappliadmin** [comment]
applying the patch on 5.76.1 leads to fatal error. Many files to be patched were lacking. Should I upgrade to RC (5.77)?

**Comment by Guydn** [comment]
I still do not see the fields that are on the screenshot.

**Comment by allinappliadmin** [comment]
> @allinappliadmin yes upgrading to the RC should help you apply the patch.

Still having a missing file notice while patching the RC
```
|diff --git a/tools/extensions/phpstorm/Civi/PhpStorm/Api4Generator.php b/tools/extensions/phpstorm/Civi/PhpStorm/Api4Generator.php
|index f5ed0e2af178..9e4e1b3f54a9 100644
|--- a/tools/extensions/phpstorm/Civi/PhpStorm/Api4Generator.php
|+++ b/tools/extensions/phpstorm/Civi/PhpStorm/Api4Generator.php
```
@Guydn can you test and confirm it works as expected?

**Comment by colemanw** [comment]
@allinappliadmin that file is not important. You can just delete it from the patch.

**Comment by Guydn** [comment]
I cannot input anything in the field "location Id" (5.77 beta1, WP 6.6.1)
CiviCRM Event location step 1
CiviCRM event location step 2
CiviCRM event location step 3

**Comment by Guydn** [comment]
On another environment (Civi 5.73, WP 6.6) managed by @mlutfy the patch works (good news !) but this patch was complex to implement.

**Comment by colemanw** [comment]
Ok great, thanks @Guydn and @mlutfy. Indeed this is a big changeset, but it includes lots of test coverage so let's get this merged so you don't have to deal with complex patches anymore!

**Comment by Guydn** [comment]
Thanks a lot @colemanw it works well now on wpmaster :-) !

**Comment by Guydn** [comment]
I have discovered a issue (Wpmaster 5.79 Alpha1) that is critical to use the feature https://lab.civicrm.org/dev/core/-/issues/5465

**Pull request merged**